File: cpp/1451.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int main()
{
    string sentence;
    while (getline(cin, sentence)) {
        string stack = "", aux = "";
        vector<string> home;
        int c = -1;
        bool flag = false;
        for (auto i: sentence) {
            if (i == '[') {
                flag = true;
                home.push_back(aux);
                c++;
            } else if (i == ']') {
                flag = false;
            } else if (flag) {
                home.at(c) += i;
            } else {
                stack += i;
            }
        }
        for (int i = home.size() - 1; i > -1; i--) {
            cout << home.at(i);
        }
        cout << stack << '\n';
    }
    return 0;
}

File: py/1451.py
def main() -> None:
    while True:
        try:
            sentence = input()
        except EOFError:
            break  
        stack = []
        home = []
        count = -1
        flag = False
        for i in sentence:
            if i == '[':
                flag = True
                home.append([])
                count += 1
            elif i == ']':
                flag = False
            elif flag:
                home[count].append(i)
            else:
                stack.append(i)
        home.reverse()
        print(*[''.join(i) for i in home], *stack, sep='')


if __name__ == '__main__':
    main()